src/Compilers/CSharp/Test/Emit2/Semantics/LockTests.cs

```diff
@@ -0,0 +1,1102 @@
+            verify: Verification.FailsILVerify);
+        verifier.VerifyDiagnostics();
+        var il = """
+            {
+              // Code size       52 (0x34)
+              .maxstack  1
+              .locals init (System.Threading.Lock.Scope V_0)
+              IL_0000:  ldstr      "1"
+              IL_0005:  call       "void System.Console.Write(string)"
+              IL_000a:  ldsfld     "System.Threading.Lock C._lock"
+              IL_000f:  callvirt   "System.Threading.Lock.Scope System.Threading.Lock.EnterLockScope()"
+              IL_0014:  stloc.0
+              .try
+              {
+                IL_0015:  ldstr      "2"
+                IL_001a:  call       "void System.Console.Write(string)"
+                IL_001f:  leave.s    IL_0029
+              }
+              finally
+              {
+                IL_0021:  ldloca.s   V_0
+                IL_0023:  call       "void System.Threading.Lock.Scope.Dispose()"
+                IL_0028:  endfinally
+              }
+              IL_0029:  ldstr      "3"
+              IL_002e:  call       "void System.Console.Write(string)"
+              IL_0033:  ret
+            }
+            """;
+        verifier.VerifyIL("C.M2", il);
+        verifier.VerifyIL("C.M1", il);
+    }
+
+    [Fact]
+    public void MissingEnterLockScope()
+    {
+        var source = """
+            System.Threading.Lock l = new();
+            lock (l) { }
+
+            namespace System.Threading
+            {
+                public class Lock { }
+            }
+            """;
+        CreateCompilation(source).VerifyEmitDiagnostics(
+            // (2,7): error CS0656: Missing compiler required member 'System.Threading.Lock.EnterLockScope'
+            // lock (l) { }
+            Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "l").WithArguments("System.Threading.Lock", "EnterLockScope").WithLocation(2, 7));
+    }
+
+    [Fact]
+    public void EnterLockScopeReturnsVoid()
+    {
+        var source = """
+            System.Threading.Lock l = new();
+            lock (l) { }
+
+            namespace System.Threading
+            {
+                public class Lock
+                {
+                    public void EnterLockScope() { }
+                }
+            }
+            """;
+        CreateCompilation(source).VerifyEmitDiagnostics(
+            // (2,7): error CS0656: Missing compiler required member 'System.Threading.Lock.EnterLockScope'
+            // lock (l) { }
+            Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "l").WithArguments("System.Threading.Lock", "EnterLockScope").WithLocation(2, 7));
+    }
+
+    [Fact]
+    public void EnterLockScopeTakesArguments()
+    {
+        var source = """
+            System.Threading.Lock l = new();
+            lock (l) { }
+
+            namespace System.Threading
+            {
+                public class Lock
+                {
+                    public Scope EnterLockScope(int arg) => new Scope();
```

Comment:
>int arg [](http://example.com/codeflow?start=48&length=7)

Please add flavors of this test:
- with an optional parameter
- with `params` parameter

Reply:
For "Dispose" as well